File: MiniMax Algorithm/minimax.js
/**
 * Calculates the best move for the current player using the Minimax algorithm with Alpha-Beta pruning.
 * @param {Array} board - An array of 9 elements representing the Tic-Tac-Toe grid.
 * Values: 'X', 'O', or null.
 * Example input: ['X', 'O', 'X', null, 'O', null, null, null, null]
 */
function getBestMoveMinimax(board) {

    /**
     * Internal helper to determine the game state.
     * Checks all winning rows, columns, and diagonals.
     */
    function checkWinner(b) {
        const lines = [
            [0, 1, 2], [3, 4, 5], [6, 7, 8], // Rows
            [0, 3, 6], [1, 4, 7], [2, 5, 8], // Columns
            [0, 4, 8], [2, 4, 6]             // Diagonals
        ];
        for (let l of lines) {
            if (b[l[0]] && b[l[0]] === b[l[1]] && b[l[0]] === b[l[2]]) {
                return { w: b[l[0]] }; // Returns 'X' or 'O' as winner
            }
        }
        return b.includes(null) ? null : { w: 'TIE' }; // Returns 'TIE' or null
    }

    /**
     * Core recursive Minimax function with Alpha-Beta pruning.
     */
    function minimax(b, depth, alpha, beta, isMaximizing) {
        let result = checkWinner(b);
        if (result) {
            if (result.w === 'O') return 10 - depth;
            if (result.w === 'X') return depth - 10;
            return 0;
        }

        if (isMaximizing) {
            let bestScore = -Infinity;
            for (let i = 0; i < 9; i++) {
                if (!b[i]) {
                    b[i] = 'O';
                    let score = minimax(b, depth + 1, alpha, beta, false);
                    b[i] = null;
                    bestScore = Math.max(score, bestScore);
                    alpha = Math.max(alpha, bestScore);
                    if (beta <= alpha) break; // Beta cut-off
                }
            }
            return bestScore;
        } else {
            let bestScore = Infinity;
            for (let i = 0; i < 9; i++) {
                if (!b[i]) {
                    b[i] = 'X';
                    let score = minimax(b, depth + 1, alpha, beta, true);
                    b[i] = null;
                    bestScore = Math.min(score, bestScore);
                    beta = Math.min(beta, bestScore);
                    if (beta <= alpha) break; // Alpha cut-off
                }
            }
            return bestScore;
        }
    }

    // Determine current player (Assumes 'X' starts first if counts are equal)
    const xCount = board.filter(v => v === 'X').length;
    const oCount = board.filter(v => v === 'O').length;
    const currentPlayer = (oCount < xCount) ? 'O' : 'X';
    const isMaximizing = (currentPlayer === 'O');

    let bestScore = isMaximizing ? -Infinity : Infinity;
    let move = -1;

    // Search for the best move index
    for (let i = 0; i < 9; i++) {
        if (!board[i]) {
            board[i] = currentPlayer;
            let score = minimax(board, 0, -Infinity, Infinity, !isMaximizing);
            board[i] = null;

            if (isMaximizing) {
                if (score > bestScore) {
                    bestScore = score;
                    move = i;
                }
            } else {
                if (score < bestScore) {
                    bestScore = score;
                    move = i;
                }
            }
        }
    }

    return { move, score: bestScore };
}

File: MiniMax Algorithm/minimax.py
import math

def check_winner(b):
    lines = [[0,1,2],[3,4,5],[6,7,8],[0,3,6],[1,4,7],[2,5,8],[0,4,8],[2,4,6]]
    for l in lines:
        if b[l[0]] and b[l[0]] == b[l[1]] == b[l[2]]:
            return b[l[0]]
    return "TIE" if None not in b else None

def minimax(b, d, a, bt, is_m):
    r = check_winner(b)
    if r == "O": return 10 - d
    if r == "X": return d - 10
    if r == "TIE": return 0
    
    if is_m:
        s = -math.inf
        for i in range(9):
            if b[i] is None:
                b[i] = "O"
                s = max(s, minimax(b, d + 1, a, bt, False))
                b[i] = None
                a = max(a, s)
                if bt <= a: break
        return s
    else:
        s = math.inf
        for i in range(9):
            if b[i] is None:
                b[i] = "X"
                s = min(s, minimax(b, d + 1, a, bt, True))
                b[i] = None
                bt = min(bt, s)
                if bt <= a: break
        return s

def get_best_move(board, player):
    best_val = -math.inf if player == "O" else math.inf
    move = -1
    for i in range(9):
        if board[i] is None:
            board[i] = player
            val = minimax(board, 0, -math.inf, math.inf, player == "X")
            board[i] = None
            if player == "O":
                if val > best_val:
                    best_val, move = val, i
            else:
                if val < best_val:
                    best_val, move = val, i
    return move

print("""
Board Index Reference:
 0 | 1 | 2 
-----------
 3 | 4 | 5 
-----------
 6 | 7 | 8 

Example Input: X, None, O, None, X, None, None, None, None
""")

raw_input = input("Enter board state (9 values separated by comma): ")
board = [val.strip() if val.strip() in ["X", "O"] else None for val in raw_input.split(",")]
current_player = input("Enter current player (X or O): ").strip().upper()

best_idx = get_best_move(board, current_player)
print(f"Best Move Index: {best_idx}")